### ISSUE TITLE
WebApplicationFactory: how to specify a proxy for all created HttpClient instances?

### ISSUE THREAD

**Issue by IanKemp** [dotnet/aspnetcore#22322]
ASP.NET Core 3.1

Specifying a proxy on a `HttpClient` instance is done by creating a `HttpMessageHandler`, but there is no way to provide a `HttpMessageHandler` to `WebApplicationFactory`. `WebApplicationFactory` does expose a `CreateDefaultClient(params DelegatingHandler[] handlers)` method but:

* `DelegatingHandler.SendAsync` can only modify responses, not requests
* creating a subclass of `DelegatingHandler` with its `InnerHandler` set to a proxied `HttpMessageHandler` has no effect because the `CreateDefaultClient` method sets up the handler chain such that all the handlers link to each other via `InnerHandler`, which means that any provided `InnerHandler` is overwritten

To make this better, there is no way to get at the default set of handlers created since `WebApplicationFactoryClientOptions.CreateHandlers()` (which is called by `WebApplicationFactory.CreateDefaultClient()`) is internal. And all of the `CreateClient`/`CreateDefaultClient` methods are non-virtual, so there's no way to override them.

What do?

**Comment by Tratcher** [comment]
Why do you want to set a proxy? This HttpClient only sends requests to the server in-memory, it never aceeses the network and can't send requests anywhere else.

**Comment by IanKemp** [comment]
> Why do you want to set a proxy? This HttpClient only sends requests to the server in-memory, it never aceeses the network and can't send requests anywhere else.

My test is calling `WebApplicationFactory<T>.CreateClient().GetAsync(url)` which is failing with a 404 when I know it should not be, so I'm trying to use Fiddler to see the actual URL that the `HttpClient` is hitting. I already tried setting the `http_proxy` and `https_proxy` environment variables, which also don't work.

But I figured it out: I had to subclass `DelegatingHandler` and override `SendAsync`, add my subclass to the collection of handlers passed to `CreateDefaultClient`, and then set a breakpoint in my handler's `SendAsync` and inspect the incoming `HttpRequestMessage`'s `RequestUri`. Still waaay more effort than it should be, but at least I can get to what I need.
